src/vm/dllimport.cpp

```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
 #endif // !FEATURE_PAL
 
 // static
-NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryFromPath(LPCWSTR libraryPath)
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryFromPath(LPCWSTR libraryPath, BOOL throwOnError)
 {
     STANDARD_VM_CONTRACT;
 
     LoadLibErrorTracker errorTracker;
     const NATIVE_LIBRARY_HANDLE hmod =
         LocalLoadLibraryHelper(libraryPath, GetLoadWithAlteredSearchPathFlag(), &errorTracker);
-    if (hmod == nullptr)
+    
+    if (throwOnError && (hmod == nullptr))
     {
         SString libraryPathSString(libraryPath);
         errorTracker.Throw(libraryPathSString);
     }
     return hmod;
 }
 
-/* static */
+// static 
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryByName(LPCWSTR libraryName, Assembly *callingAssembly, 
+                                                 DWORD dllImportSearchPathFlag, BOOL throwOnError)
+{
+    STANDARD_VM_CONTRACT;
+
+    LoadLibErrorTracker errorTracker;
+
+    BOOL searchAssemblyDirectory = dllImportSearchPathFlag & DLLIMPORTSEARCHPATH_ASSEMBLYDIRECTORY;
+    dllImportSearchPathFlag &= ~DLLIMPORTSEARCHPATH_ASSEMBLYDIRECTORY;
+
+    const NATIVE_LIBRARY_HANDLE hmod = 
+        LoadLibraryModuleBySearch(callingAssembly, searchAssemblyDirectory, dllImportSearchPathFlag, &errorTracker, libraryName);
+
+    if (throwOnError && (hmod == nullptr))
+    {
+        SString libraryPathSString(libraryName);
+        errorTracker.Throw(libraryPathSString);
+    }
+
+    return hmod;
+}
+
+// static
+void NDirect::FreeNativeLibrary(NATIVE_LIBRARY_HANDLE handle)
+{
+    STANDARD_VM_CONTRACT;
+
+    if (handle == NULL)
+        COMPlusThrowArgumentNull(W("handle"), W("Arg_InvalidHandle"));
+
+#ifndef FEATURE_PAL
+    BOOL retVal = FreeLibrary(handle);
+#else // !FEATURE_PAL
+    BOOL retVal = PAL_FreeLibraryDirect(handle);
+#endif // !FEATURE_PAL
+
+    if (retVal == 0)
```

Comment:
I thought we decided to not throw if we failed to free the library? Same for the handle being null? (i.e. Free should never throw). I could be mis-remembering.

Reply:
@AaronRobinsonMSFT  and I talked about this. FreeLibrary will throw if the close operation fails (following the principle that dlclose and FreeLibrary() return error codes).
However, it won't throw on null input, to avoid further null propagation (ex: in finally blocks).